### PR TITLE
Implement co-broadcasting in operator overloading

### PR DESCRIPTION
**Pull request by SparrowLii**
This PR did the following to implement co_broadcasting in operator overloading
1. Implement DimMax trait, which uses the same broadcast mechanism as Numpy, for all Dimension so as to get the returned array type in operator overloading. 
2. Use map_collect() and map_collect_owned() in operator overloading to avoid redundant array traversal.

### CODE REVIEW

**Comment by bluss** [comment]
Co-broadcasting is cool and interesting! I can see that you are attacking our fixme's in code one by one and that's gratifying, but it doesn't always tell the whole story of what might be the most important issues to solve. (*)

I think the type system is still limited in what can be expressed, and we seem to have a typenum-like problem here (not really the fault of typenum, just the limits of the type system). Expressing one generic arithmetic operation is ok.. but what about two? will it be usable for the user?  Imagine this:

```rust
type Mat<D> = Array<f64, D>;
fn foo<D>(x: Mat<D>, y: Mat<D>, z: Mat<D>) -> Mat<D> {
   x + &y * &z
}
```

How would we write the actual return type of the function? Maybe in some cases this makes it harder for the user of ndarray, or can we find a great solution?

It could be that co-broadcasting should be a feature that ndarray implements, without using it by default in the arithmetic ops.

(*) Specifically see the issue about #699 - there are lots of things we want to spend the - implementation, complexity, etc - budget on in arithmetic ops.

File: src/impl_ops.rs [hunk 1]
```diff
@@ -79,27 +104,34 @@ where
...
+/// If their shapes disagree, `self` is broadcast to their broadcast shape,
+/// cloning the data if needed.
 ///
 /// **Panics** if broadcasting isn’t possible.
 impl<'a, A, B, S, S2, D, E> $trt<&'a ArrayBase<S2, E>> for ArrayBase<S, D>
 where
     A: Clone + $trt<B, Output=A>,
     B: Clone,
-    S: DataOwned<Elem=A> + DataMut,
+    S: Data<Elem=A>,
```
Comment (jturner314) [inline src/impl_ops.rs:117]: This change is unfortunate. It would be nice to continue to reuse owned allocations when possible.

File: src/numeric/impl_numeric.rs [hunk 2]
```diff
@@ -283,10 +283,11 @@ where
     ///     a.mean_axis(Axis(0)).unwrap().mean_axis(Axis(0)).unwrap() == aview0(&3.5)
     /// );
     /// ```
-    pub fn mean_axis(&self, axis: Axis) -> Option<Array<A, D::Smaller>>
+    pub fn mean_axis(&self, axis: Axis) -> Option<Array<A, <D::Smaller as BroadcastShape<Ix0>>::BroadcastOutput>>
     where
         A: Clone + Zero + FromPrimitive + Add<Output = A> + Div<Output = A>,
         D: RemoveAxis,
+        D::Smaller: BroadcastShape<Ix0>,
```
[the remark this answers is not in the document]
Comment (jturner314) [reply src/numeric/impl_numeric.rs:290]: This is a good point and something I hadn't considered before. Unfortunately, I don't think there's a way to avoid the extra bound without improvements to the language (either specialization or enum-variants-as-types), since the compiler needs to know that `BroadcastShape` is implemented for all possible combinations of dimension types.

[the remark this answers is not in the document]
Comment (bluss) [reply src/numeric/impl_numeric.rs:290]: Let's see how far we get with the blanket implementation for all `Dimension` with output `Self` and then maybe it can be added that `Ix0` can broadcast to any other dimension.

[the remark this answers is not in the document]
Comment (SparrowLii) [reply src/numeric/impl_numeric.rs:290]: I can't think of a good way :( If we add a statement like this:

        impl<D: Dimension> BroadcastShape<Ix0> for D {
            type BroadcastOutput = D;
        }
Then it will conflict with the statement on broadcast.rs:62( Although they don’t actually conflict ).

**Comment by SparrowLii** [comment]
> How would we write the actual return type of the function? Maybe in some cases this change makes it harder for the user of ndarray, or can we find a great solution?
If the dimension types of the parameters are all D, the result can also be written directly as D. But if the dimensions are not the same, it will be really troublesome. That's right, if the cost of using operators is that additional restrictions must be added, it maybe better to risk panic. I'm sorry I can't think of a better way for the time being. Thanks for pointing out.

File: src/impl_ops.rs (hunk 1 above)
Comment (SparrowLii) [reply src/impl_ops.rs:117]: Thanks for pointing out

**Comment by bluss** [comment]
Changed the title from "fix" since it's a new feature, not a fix. There is more feedback and thoughts to be given, when I have time. 🙂

**Comment by bluss** [comment]
Here's a trick that actually works to resolve the `mean` situation. Just to show you a neat trick to use. Ignore the `broadcast_shape` method change for now - that one can be fixed later (debatable if it should be a method at all).

What we do here is that we put the `BroadcastShape<Self>` and `BroadcastShape<Ix0>` requirements into requirements of the `Dimension` itself. That way we rearrange the situation so that instead every type that implements Dimension has to prove - for itself - that it implements these two broadcasts. It seems to sidestep the overlapping problem.

See this temporary branch for that diff: https://github.com/rust-ndarray/ndarray/pull/new/bluss/pr-898-broadcast-shape

**Comment by jturner314** [comment]
> Here's a trick that actually works to resolve the `mean` situation.

That's a really nice solution. Is it possible to add `BroadcastShape<Self::Smaller, BroadcastOutput=Self>` and `BroadcastShape<Self::Larger, BroadcastOutput=Self::Larger>` bounds too? If so, I would think all four bounds would avoid a need for users to add `BroadcastShape` bounds to their functions in most cases.

**Comment by SparrowLii** [comment]
> What we do here is that we put the BroadcastShape<Self> and BroadcastShape<Ix0> requirements into requirements of the Dimension itself. That way we rearrange the situation so that instead every type that implements Dimension has to prove - for itself - that it implements these two broadcasts. It seems to sidestep the overlapping problem.

That is amazing! It's so surprising it works! However, we still can’t add statements like BroadcastShape<Self::Smaller, BroadcastOutput=Self> because it will cause loops in analysis passes. But the exciting thing is that I tried this way in the definition of Dimension:

    /// Next smaller dimension (if applicable)
    type Smaller: Dimension + BroadcastShape<Self, BroadcastOutput=Self>;
    /// Next larger dimension
    type Larger: Dimension + RemoveAxis + BroadcastShape<Self, BroadcastOutput=Self::Larger>;

And it did run successfully in my nightly version (cargo 1.49.0-nightly (d5556aeb8 2020-11-04) ), I think we can achieve this amazing work as soon as the 1.49 version is stable.

**Comment by bluss** [comment]
Note that `+ BroadcastShape<Self::Smaller, BroadcastOutput=Self>` does work it just has to be written like `+ BroadcastShape<<Self as Dimension>::Smaller, BroadcastOutput=Self>` to help the compiler.

**Comment by jturner314** [comment]
We could also add this bound to `RemoveAxis`: `BroadcastShape<<<Self as Dimension>::Smaller as Dimension>::Larger, BroadcastOutput = Self>`.

File: src/impl_ops.rs [hunk 3]
```diff
@@ -53,24 +54,48 @@ macro_rules! impl_binary_op(
...
     S2: Data<Elem=B>,
-    D: Dimension,
+    D: Dimension + BroadcastShape<E>,
     E: Dimension,
 {
-    type Output = ArrayBase<S, D>;
-    fn $mth(self, rhs: ArrayBase<S2, E>) -> ArrayBase<S, D>
+    type Output = Array<A, <D as BroadcastShape<E>>::BroadcastOutput>;
```
Comment (bluss) [inline src/impl_ops.rs:72]: A feature is removed here - the ability to have arithmetic that preserves ArcArray (ArcArray in and out).

File: src/impl_ops.rs [hunk 4]
```diff
@@ -53,24 +54,48 @@ macro_rules! impl_binary_op(
...
+where
+    A: Clone + $trt<B, Output=A>,
+    B: Clone,
+    S: Data<Elem=A>,
+    S2: Data<Elem=B>,
+    D: Dimension + BroadcastShape<E>,
+    E: Dimension,
+{
+    type Output = Array<A, <D as BroadcastShape<E>>::BroadcastOutput>;
+    fn $mth(self, rhs: ArrayBase<S2, E>) -> Self::Output
```
Comment (bluss) [inline src/impl_ops.rs:98]: If this impl this method should aim to use rhs' allocation

File: src/impl_ops.rs (hunk 1 above)
Comment (bluss) [reply src/impl_ops.rs:117]: They are reused still in some cases

File: src/impl_ops.rs [hunk 5]
```diff
@@ -79,27 +104,34 @@ where
...
-    S: DataOwned<Elem=A> + DataMut,
+    S: Data<Elem=A>,
     S2: Data<Elem=B>,
-    D: Dimension,
+    D: Dimension + BroadcastShape<E>,
     E: Dimension,
 {
-    type Output = ArrayBase<S, D>;
-    fn $mth(mut self, rhs: &ArrayBase<S2, E>) -> ArrayBase<S, D>
+    type Output = Array<A, <D as BroadcastShape<E>>::BroadcastOutput>;
```
Comment (bluss) [inline src/impl_ops.rs:122]: We'd like to preserve the type of owned array here

File: src/impl_ops.rs [hunk 6]
```diff
@@ -79,27 +104,34 @@ where
...
+    fn $mth(self, rhs: &ArrayBase<S2, E>) -> Self::Output
     {
-        self.zip_mut_with(rhs, |x, y| {
+        let shape = self.dim.broadcast_shape(&rhs.dim).unwrap();
+        let mut self_ = if shape.slice() == self.dim.slice() {
+            self.into_owned().into_dimensionality::<<D as BroadcastShape<E>>::BroadcastOutput>().unwrap()
```
Comment (bluss) [inline src/impl_ops.rs:127]: The method body is quite messy but it's not the highest priority to solve first.

File: src/impl_ops.rs [hunk 7]
```diff
@@ -117,13 +150,21 @@ where
...
     S: Data<Elem=A>,
     S2: Data<Elem=B>,
-    D: Dimension,
+    D: Dimension + BroadcastShape<E>,
     E: Dimension,
 {
-    type Output = Array<A, D>;
-    fn $mth(self, rhs: &'a ArrayBase<S2, E>) -> Array<A, D> {
-        // FIXME: Can we co-broadcast arrays here? And how?
-        self.to_owned().$mth(rhs)
```
Comment (bluss) [inline src/impl_ops.rs:126]: It needs to be mentioned that the number one thing I'd want to fix with this impl is not co-broadcasting, but removing the `.to_owned()` call - that is we don't want to traverse the lhs array twice (once to copy and once to do the arithmetic).

File: src/impl_ops.rs [hunk 8]
```diff
@@ -117,13 +150,21 @@ where
...
+    type Output = Array<A, <D as BroadcastShape<E>>::BroadcastOutput>;
+    fn $mth(self, rhs: &'a ArrayBase<S2, E>) -> Self::Output {
+        let shape = self.dim.broadcast_shape(&rhs.dim).unwrap();
+        let mut self_ = if shape.slice() == self.dim.slice() {
+            self.to_owned().into_dimensionality::<<D as BroadcastShape<E>>::BroadcastOutput>().unwrap()
+        } else {
+            self.broadcast(shape).unwrap().to_owned()
```
Comment (bluss) [inline src/impl_ops.rs:162]: This line of code uses `to_owned` on a broadcasted array - so we copy a large array (separately from the arithmetic) - ideally we should not do this - allocate the destination array uninitialized (`maybe_uninit`) and broadcast both inputs as views and do the operation.

File: tests/array.rs [hunk 9]
```diff
@@ -485,7 +485,7 @@ fn test_add() {
     }
 
     let B = A.clone();
-    A = A + &B;
+    let A = A + &B;
```
Comment (bluss) [inline tests/array.rs:488]: It's a goal that we can revert this change 🙂

File: tests/array.rs [hunk 10]
```diff
@@ -1558,6 +1558,53 @@ fn insert_axis_view() {
...
+        &Dim(vec![1, 1, 3, 1, 5, 1, 7]),
+        &Dim([2, 1, 4, 1, 6, 1]),
+        Ok(Dim(IxDynImpl::from(v.as_slice()))),
+    );
+    let d = Dim([1, 2, 1, 3]);
+    test_co(&d, &d, Ok(d));
+    test_co(
+        &Dim([2, 1, 2]).into_dyn(),
+        &Dim(0),
+        Ok(Dim([2, 1, 2]).into_dyn()),
```
Comment (bluss) [inline tests/array.rs:1592]: Should probably be a shape error when one the rhs is this zero dimension?

```python
>>> import numpy as np
>>> np.arange(4.)
array([0., 1., 2., 3.])
>>> a = np.arange(4.).reshape((2, 1, 2))
>>> a
array([[[0., 1.]],

       [[2., 3.]]])
>>> b = a[:0, :, :0]
>>> b = b.reshape((0, ))
>>> b
array([], dtype=float64)
>>> b.shape
(0,)
>>> a + b
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: operands could not be broadcast together with shapes (2,1,2) (0,)
>>>
```

**Review by bluss — COMMENTED** [review]
I'm sorry I haven't at this time provided a high-level plan

**Comment by SparrowLii** [comment]
I used uninitialized() to create an uninitialized array, and added the zip_mut_from_pair method to impl_methods.rs to get the result. Now no matter whether we need co_broadcast or not, we only need to traverse self and rhs once each. And ArcArray will not be turned into Array. But there is still a problem: both uninitialized() and maybe_uninit() methods require elements to implement the copy trait (I don’t know the reason behind this). So it is still unavoidable to change the signature of the method in impl_numeric.rs.

File: src/impl_ops.rs (hunk 4 above)
Comment (SparrowLii) [reply src/impl_ops.rs:102]: Thanks for correcting. I made a separate implementation of it now.

File: src/impl_ops.rs (hunk 3 above)
Comment (SparrowLii) [reply src/impl_ops.rs]: That's right. We should not abandon the limitation of DataOwned. For non-DataOwned arrays, you can use &.

File: tests/array.rs (hunk 9 above)
Comment (SparrowLii) [reply tests/array.rs]: That's right. We can achieve it by keeping the limit of DataOwned.

File: tests/array.rs (hunk 10 above)
Comment (SparrowLii) [reply tests/array.rs:1592]: Yes. I did not test the zero dimension in numpy. I will test them and be consistent with numpy.

**Comment by bluss** [comment]
uninitialized is hard to use correctly. I haven't removed it yet because I think I have managed to put it on sound ground - it *is possible* to use it soundly.

Unfortunately this method

1. Has some usage by actual users (github code search confirms)
2. This usage is from casual inspection almost always technically incorrect

Please use `maybe_uninit`. 🙂 If you do, then I don't have to go through and point out where your usage of `uninitialized` breaks the soundness rules 🙂

File: src/dimension/broadcast.rs [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+                return Err(from_kind(ErrorKind::IncompatibleShape));
+            }
+        }
+    }
+    Ok(out)
+}
+
+pub trait BroadcastShape<Other: Dimension> {
+    /// The resulting dimension type after broadcasting.
+    type BroadcastOutput: Dimension;
```
Comment (bluss) [inline src/dimension/broadcast.rs:46]: A trait with an "output" type (See RawDataSubst and FnOnce as inspiration) should almost always prefer the associated type name "Output" as a matter of convention. Think of it as a type-level function's return value.

File: src/impl_methods.rs [hunk 12]
```diff
@@ -1968,6 +1968,46 @@ where
...
+    /// Traverse two arrays in unspecified order, in lock step,
+    /// calling the closure `f` on each element pair, and put
+    /// the result into the corresponding element of self.
+    pub fn zip_mut_from_pair<B, C, S1, S2, F>(&mut self, lhs: &ArrayBase<S1, D>, rhs: &ArrayBase<S2, D>, f: F, )
+    where
+        S: DataMut,
+        S1: Data<Elem = B>,
+        S2: Data<Elem = C>,
+        F: Fn(&B, &C) -> A,
+    {
```
Comment (bluss) [inline src/impl_methods.rs:1980]: Why not just use `Zip` in this method? `Zip::map_assign_into` should do it (assign into self)

File: src/impl_ops.rs [hunk 13]
```diff
@@ -108,22 +175,33 @@ where
...
-        // FIXME: Can we co-broadcast arrays here? And how?
-        self.to_owned().$mth(rhs)
+    type Output = Array<A, <D as BroadcastShape<E>>::BroadcastOutput>;
+    fn $mth(self, rhs: &'a ArrayBase<S2, E>) -> Self::Output {
+        let shape = self.dim.broadcast_shape(&rhs.dim).unwrap();
+        // SAFETY: Overwrite all the elements in the array after
+        // it is created via `zip_mut_from_pair`.
+        let mut out = unsafe {
+            Self::Output::uninitialized(shape.clone().into_pattern())
+        };
```
Comment (bluss) [inline src/impl_ops.rs:198]: On closer thought, it would be better to just use plainly `map_collect` here - it allocates the output array for you, in an ok memory layout and you don't have to write the uninit logic correctly yourself.

**Comment by SparrowLii** [comment]
> Please use maybe_uninit that I mentioned. 🙂 If you do, then I don't have to go through and point out where your usage of uninitialized breaks the soundness rules 🙂

I tried to use maybe_uninit(), but I can’t think of a simple way to generate an ArrayBase with the data type MaybeUninit<A>. So I added a feature called MaybeUninitSubst to get the mapping from ArrayBase with element A to MaybeUnint<A>. In addition used Zip to get the results.

File: src/dimension/broadcast.rs (hunk 11 above)
Comment (SparrowLii) [reply src/dimension/broadcast.rs]: Thanks much for teaching.  I changed BroadCastOutput to Output

File: src/impl_methods.rs (hunk 12 above)
Comment (SparrowLii) [reply src/impl_methods.rs]: Right, we can use Zip. This function does not need to be added.

**Comment by bluss** [comment]
Good fighting. The type related problems with the maybe_uninit constructor are not so fun, we're really feeling that we miss the "GAT" feature here (still experimental).

File: tests/array.rs (hunk 10 above)
Comment (SparrowLii) [reply tests/array.rs]: As for zero dimension in numpy, if we do like this:

    import numpy as np
    a = np.arange(4.).reshape((2, 2, 1))
    print(a.shape)
    
    b = a[:2,:2,:0]
    b = b.reshape((0))
    print(b.shape)
    
    print(a+b)

It's just fine:

    (2, 2, 1)
    (0,)
    []

In 1 and 0, numpy will also choose 0：

    import numpy as np
    a = np.arange(2.).reshape((1, 2, 1))
    print(a.shape)
    
    b = a[:0,:2,:1]
    b = b.reshape((0,2,1))
    print(b.shape)
    c=a+b
    print(c.shape)

Output:

    (1, 2, 1)
    (0, 2, 1)
    (0, 2, 1)

Then we can look PyArray_BroadcastToShape method in numpy/numpy/core/src/multiarray/iterators:204:

        for (i = 0; i < PyArray_NDIM(ao); i++, j++) {
            if (PyArray_DIMS(ao)[i] == 1) {
                continue;
            }
            if (PyArray_DIMS(ao)[i] != dims[j]) {
                compat = 0;
                break;
            }
        }
        if (!compat) {
            goto err;
        }

numpy treats 0 values just like other values greater than 1.

**Comment by bluss** [comment]
I'm working on fixing the maybe uninit functionality so that it can do what we need.

**Comment by SparrowLii** [comment]
> I'm working on fixing the maybe uninit functionality so that it can do what we need.

Cool! It will be very exciting if it and co_broadcast both land.

**Comment by SparrowLii** [comment]
@bluss Thank you so much for the amazing work! I have rebased the pr and used map_collect_owned according to your guide to avoid unsafe code.

File: src/dimension/broadcast.rs [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+    let mut out = Output::zeros(shape1.ndim());
+    let out_slice = out.slice_mut();
+    let s1 = shape1.slice();
+    let s2 = shape2.slice();
+    // Uses the [NumPy broadcasting rules]
+    // (https://docs.scipy.org/doc/numpy/user/basics.broadcasting.html#general-broadcasting-rules).
+    //
+    // Zero dimension element is not in the original rules of broadcasting.
+    // We currently treat it like any other number greater than 1. As numpy does.
+    for i in 0..shape1.ndim() {
```
Comment (bluss) [inline src/dimension/broadcast.rs:27]: index loop that should be a `zip(&mut out_slice, &s1)` loop.

File: src/dimension/broadcast.rs [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+        out_slice[i] = s1[i];
+    }
+    for i in 0..shape2.ndim() {
```
Comment (bluss) [inline src/dimension/broadcast.rs:30]: index loop that should be a `zip(&mut out_slice[k..], &s2)` loop

Comment (SparrowLii) [reply src/dimension/broadcast.rs:30]: OK. Thanks a lot for taking the time to review.

File: src/dimension/broadcast.rs [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+    // Swap the order if d2 is longer.
+    if overflow {
+        return broadcast_shape::<D2, D1, Output>(shape2, shape1);
+    }
+    // The output should be the same length as shape1.
+    let mut out = Output::zeros(shape1.ndim());
+    // Uses the [NumPy broadcasting rules]
+    // (https://docs.scipy.org/doc/numpy/user/basics.broadcasting.html#general-broadcasting-rules).
+    //
+    // Zero dimension element is not in the original rules of broadcasting.
```
Comment (bluss) [inline src/dimension/broadcast.rs:22]: Which original rules? `ArrayBase::broadcast` seems concsistent with these rules already. Also, we need to ensure here that we do exactly the same thing as in `ArrayBase::broadcast`.

**Review by bluss — CHANGES_REQUESTED** [review]
It's starting to getting in shape, we've managed to reduce the arithmetic ops trait requirements to a minimum (optimal would be to have all dimensions implement `BroadcastShape<E>` but it seems impossible and this is acceptable for now).

Arithmetic ops and broadcasting docs (on ArrayBase) look like they need an update

File: tests/array.rs [hunk 17]
```diff
@@ -1566,6 +1613,49 @@ fn arithmetic_broadcast() {
...
+
+    // shared array
+    let sa = a.to_shared();
+    let sa2 = sa.to_shared();
+    let sb = b.to_shared();
+    let sb2 = sb.to_shared();
+    let sc = c.to_shared();
+    let sc2 = sc.into_shared();
+    assert_eq!(
+        sa2 + &sb2 + sc2.into_owned(),
```
[the remark this answers is not in the document]
Comment (SparrowLii) [reply tests/array.rs:1609]: Yes, just for mixed operations

File: src/impl_ops.rs [hunk 18]
```diff
@@ -108,22 +160,25 @@ where
...
 {
-    type Output = Array<A, D>;
-    fn $mth(self, rhs: &'a ArrayBase<S2, E>) -> Array<A, D> {
-        // FIXME: Can we co-broadcast arrays here? And how?
-        self.to_owned().$mth(rhs)
+    type Output = Array<A, <D as BroadcastShape<E>>::Output>;
+    fn $mth(self, rhs: &'a ArrayBase<S2, E>) -> Self::Output {
+        let shape = self.dim.broadcast_shape(&rhs.dim).unwrap();
+        let lhs = self.broadcast(shape.clone()).unwrap();
+        let rhs = rhs.broadcast(shape).unwrap();
```
[the remark this answers is not in the document]
Comment (SparrowLii) [reply src/impl_ops.rs:177]: That's right. But the method to convert the dim of ArrayBase seems to have only into_dimensionality and broadcast. However, the former cannot be used for reference types. I think we can add the judgment whether the shape is the same in the broadcast.

File: src/dimension/broadcast.rs (hunk 16 above)
Comment (SparrowLii) [reply src/dimension/broadcast.rs]: I did not find a description of the zero value in the Numpy broadcast rules cited above. However, if there is no difference between a zero value and a value greater than 1, I think this explanation can be omitted. 
`ArrayBase::broadcast` only detects the shape mismatch, and does not make any changes to dim.

File: src/impl_ops.rs [hunk 19]
```diff
@@ -90,16 +96,59 @@ where
...
-    type Output = ArrayBase<S, D>;
-    fn $mth(mut self, rhs: &ArrayBase<S2, E>) -> ArrayBase<S, D>
+    type Output = ArrayBase<S, <D as BroadcastShape<E>>::Output>;
+    fn $mth(self, rhs: &ArrayBase<S2, E>) -> Self::Output
     {
-        self.zip_mut_with(rhs, |x, y| {
-            *x = x.clone() $operator y.clone();
-        });
-        self
+        if self.ndim() == rhs.ndim() && self.shape() == rhs.shape() {
```
Comment (jturner314) [inline src/impl_ops.rs:105]: I haven't reviewed the PR in detail, but I just wanted to make a small comment here. There's another case that's worth considering: when it's possible to broadcast the RHS to match the existing shape of the LHS. In this case, it's possible to reuse the existing LHS allocation (like in the current release of `ndarray` and the `if` branch here), instead of collecting into a new array (like the `else` branch here).

This observation also applies analogously to the other ops impl with `DataOwned`.

File: src/dimension/broadcast.rs [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+                out_slice[i + k] = s2[i]
+            } else if s2[i] != 1 {
+                return Err(from_kind(ErrorKind::IncompatibleShape));
+            }
+        }
+    }
+    Ok(out)
+}
+
+pub trait BroadcastShape<Other: Dimension> {
```
[the remark this answers is not in the document]
Comment (jturner314) [reply src/dimension/broadcast.rs:37]: I don't know the answer to the tuple question, but I had another thought about this trait -- we may want to name it `DimMax`. We're going to add `DimAdd` in #570, and we'll want `DimSub` for slicing with ellipses. The name `DimMax` would be more consistent with `DimAdd`/`DimSub`.

[the remark this answers is not in the document]
Comment (bluss) [reply src/dimension/broadcast.rs:37]: ok, that's good. Can be fixed after this PR is merged, so not blocking, but we'll use that name.

File: src/dimension/broadcast.rs [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+    /// The resulting dimension type after broadcasting.
+    type Output: Dimension;
+
+    /// Determines the shape after broadcasting the dimensions together.
```
[the remark this answers is not in the document]
Comment (bluss) [reply src/dimension/broadcast.rs:41]:
```suggestion
    /// Determines the shape after broadcasting the shapes together.
```

**Review by bluss — COMMENTED** [review]
The PR is basically ready, a few minor things we want to change and then we merge. A few of the other issues mentioned can be fixed after merging.

File: src/impl_ops.rs (hunk 19 above)
Comment (bluss) [reply src/impl_ops.rs:105]: This is room for future improvement.

**Review by bluss — APPROVED** [review]

File: src/impl_ops.rs [hunk 22]
```diff
@@ -79,27 +83,75 @@ where
...
+            let mut out = self.into_dimensionality::<<D as BroadcastShape<E>>::Output>().unwrap();
+            out.zip_mut_with_same_shape(rhs, clone_iopf(A::$mth));
+            out
+        } else {
+            let shape = self.dim.broadcast_shape(&rhs.dim).unwrap();
+            let lhs = self.broadcast(shape.clone()).unwrap();
+            let rhs = rhs.broadcast(shape).unwrap();
```
[the remark this answers is not in the document]
Comment (SparrowLii) [reply src/impl_ops.rs]: @bluss It does make sense. We can't directly use `(&self).$mth(rhs)`, which will limit the return type to Array. I think it is a good choice to write it separately as a function.

**Comment by SparrowLii** [comment]
> The PR is basically ready, a few minor things we want to change and then we merge. A few of the other issues mentioned can be fixed after merging.

Thank you very much for reviewing. It does have points that need improvement. I added a function called broadcast_with, but if this is not the focus at the moment, maybe we can consider it in the future.

**Comment by bluss** [comment]
Ok that's good, please make broadcast_with `pub(crate)` for now - otherwise we have to design name, signature and doc comment more. Let's get to the merge.

**Comment by bluss** [comment]
I'll just make a list of "known issues" that we can fix **after** the PR review and merge

- Rename BroadcastShape to `DimMax` - we also move the broadcast method out of the trait (it can just be a free function)
- It is a wish that the `D: Dimension + BroadcastShape<E>` bound in arith ops becomes just `D: Dimension`, i.e. the DimMax with associated type becomes implied by `D: Dimension`. At this point it is not certain how this can be done.
- Preserve the allocation of the input array in arithmetic ops in more cases (when only the left or right side needs broadcast)
- Improve performance of "no-broadcasting-needed" scenario in &array + &array operation
- Potentially provide public function or method similar to `broadcast_with` - that does the broadcasting of two arrays together.

File: src/impl_methods.rs [hunk 23]
```diff
@@ -1707,6 +1707,36 @@ where
         unsafe { Some(ArrayView::new(self.ptr, dim, broadcast_strides)) }
     }
 
+    /// Calculate the views of two ArrayBases after broadcasting each other, if possible.
```
Comment (bluss) [inline src/impl_methods.rs:1710]: It will be an internal method so it doesn't matter so much, but we don't use this terminology ("ArrayBases") in the docs.
```suggestion
    /// For two arrays or views, find their common shape if possible and broadcast them as array views into that shape
```

**Comment by SparrowLii** [comment]
@bluss It has been modified accordingly.  I very much look forward to our ability to solve the above problems in the future. Especially removing the restriction like BroadCast<E>, it may represent a new language feature IMO, which will be so cool and exciting.

**Comment by bluss** [comment]
I've rebased this to fix the conflict (sorry about that). Could we update the PR description for the current state of the PR before merge?

**Comment by SparrowLii** [comment]
> I've rebased this to fix the conflict (sorry about that). Could we update the PR description for the current state of the PR before merge?

OK, it has been modified accordingly

**Comment by bluss** [comment]
Thanks :)
